src/checker/zx/ZXChecker.cpp

```diff
@@ -0,0 +1,147 @@
+        zx::ZXDiagram dPrime = zx::FunctionalityConstruction::buildFunctionality(&qc2);
+
+        if (qc1.getNancillae() != 0 || qc2.getNancillae() != 0)
+            ancilla = true;
+
+        auto p1 = invertPermutations(qc1);
+        auto p2 = invertPermutations(qc2);
+
+        // fix ancillaries to |0>
+        for (auto anc = static_cast<dd::Qubit>(qc1.getNqubits() - 1); anc >= qc1.getNqubitsWithoutAncillae(); --anc) {
+            miter.makeAncilla(anc, p1.at(anc));
+            dPrime.makeAncilla(anc, p2.at(anc));
+        }
+        miter.invert();
+        miter.concat(dPrime);
+    }
+
+    EquivalenceCriterion ZXEquivalenceChecker::run() {
+        const auto start = std::chrono::steady_clock::now();
+
+        zx::fullReduceApproximate(miter, tolerance);
+
+        bool equivalent = true;
+
+        if (miter.getNEdges() == miter.getNQubits()) {
+            const qc::Permutation  p1 = invert(invertPermutations(qc1));
+            const qc::Permutation& p2 = invert(invertPermutations(qc2));
+
+            for (std::size_t i = 0; i < miter.getNQubits(); ++i) {
+                auto& in  = miter.getInputs()[i];
+                auto& out = miter.incidentEdges(in)[0].to;
+
+                if (p1.at(static_cast<dd::Qubit>(miter.getVData(in).value().qubit)) != p2.at(static_cast<dd::Qubit>(miter.getVData(out).value().qubit))) {
+                    equivalent = false;
+                    break;
+                }
+            }
+        } else {
+            equivalent = false;
+        }
+
+        const auto end = std::chrono::steady_clock::now();
+        runtime += std::chrono::duration<double>(end - start).count();
+
+        // non-equivalence might be due to incorrect assumption about the state of ancillaries, so no information can be given
+        if (!equivalent && ancilla)
+            equivalence = EquivalenceCriterion::NoInformation;
+        else
+            equivalence = equivalent ? EquivalenceCriterion::EquivalentUpToGlobalPhase : EquivalenceCriterion::ProbablyNotEquivalent;
+
+        return equivalence;
+    }
+
+    qc::Permutation invert(const qc::Permutation& p) {
+        qc::Permutation pInv{};
+        for (auto [v, w]: p) {
+            pInv[w] = v;
+        }
+        return pInv;
+    }
+    qc::Permutation concat(const qc::Permutation& p1, const qc::Permutation& p2) { //p2 after p1
+        qc::Permutation pComb{};
+
+        for (auto [v, w]: p1) {
+            if (p2.find(w) != p2.end())
+                pComb[v] = p2.at(w);
+        }
+        return pComb;
+    }
+
+    qc::Permutation complete(const qc::Permutation& p, dd::Qubit n) {
+        qc::Permutation pComp{};
+
+        std::vector<bool> mappedTo(n, false);
+        std::vector<bool> mappedFrom(n, false);
+        for (auto [k, v]: p) {
+            mappedFrom[k] = true;
+            mappedTo[v]   = true;
+            pComp[k]      = v;
```

Comment:
wouldn't it be more straight forward to just initialize `pComb` with `p` outside the loop?